evolver/device.py

```
@@ -53,6 +66,14 @@ def validate_vials(cls, value: list[int], info: ValidationInfo):
     def __init__(self, *args, **kwargs):
         self.last_read = defaultdict(lambda: int(-1))
         super().__init__(*args, evolver=self, **kwargs)
+        # We have to turn experiment controllers passed in as configdescriptors
+        # to objects while passing in self so controllers can consume hardware
+        # and read from history, etc.
+        for experiment in self.experiments.values():
+            for i in range(len(experiment.controllers)):
+                elem = experiment.controllers[i]
+                if isinstance(elem, ConfigDescriptor):
```

Is this reflected in the type definition `experiments: dict[str, Experiment] = {}` like is `Experiment` typed somewhere as a ConfigDescriptor? Line of reasoning, if at somepoint `elem` here stops being an instance of ConfigDescriptor this condition will silently fail (i maybe wrong in the silent part) unless some typing also accompanies the condition....

---

this is from def of experiment.controllers:

```py
class Experiment(BaseInterface.Config):
    enabled: bool = True
    controllers: list[ConfigDescriptor | Controller] = []
```

It is expected to possibly be a configdescriptor. I don't think there is a case for silent failure, if a proper model instance is passed, pydantic will do the validation, otherwise the object will either have to implement the `Controller` API, or there will be an exception because it doesn't, which is what we expect.